ARROW-17181: [Docs][Python] Scalar UDF Experimental Documentation (#13687)
apache/arrow: docs/source/python/compute.rst

:class:`.Dataset` currently can be filtered using :meth:`.Dataset.to_table` method
passing a ``filter`` argument. See :ref:`py-filter-dataset` in Dataset documentation.

User-Defined Functions
======================

.. warning::
   This API is **experimental**.

>    Also, only scalar functions can currently be user-defined.

Author: I added a section for this too. Could you please check it?

   ...        "summary": "Calculate y = mx + c with Python",
   ...        "description":
   ...            "Compute the affine function y = mx + c.\n"
   ...            "This function takes three inputs, m, x and c, in order."
   ... }
   >>> input_types = {
   ...    "m" : pa.float64(),
   ...    "x" : pa.float64(),
   ...    "c" : pa.float64(),
   ... }
   >>> output_type = pa.float64()
   >>> 
   >>> def affine_with_python(ctx, m, x, c):
   ...     m = m[0].as_py()
   ...     x = x[0].as_py()
   ...     c = c[0].as_py()
   ...     return pa.array([m * x + c])
   ... 
   >>> pc.register_scalar_function(affine_with_python,
   ...                             function_name,
   ...                         function_docs,
   ...                             input_types,
   ...                             output_type)
   >>> 
   >>> pc.call_function(function_name, [pa.scalar(10.1), pa.scalar(10.2), pa.scalar(20.2)])
   <pyarrow.DoubleScalar: 123.22>

In case of all scalar inputs, make sure to return the final output as an array.

More generally, UDFs can be used with tabular data by using `dataset` API and apply a UDF function on a
dataset.

.. code-block:: python

>    >>> import pyarrow.dataset as ds

Author: @westonpace I will modify the dataset and fit it with a proper scenario.

>    >>> import pyarrow.dataset as ds

Author: I added a section about this. Could you please check if it is accurate?

                               function_name,
                               function_docs,
                               input_types,
                               output_type)

The implementation of a user-defined function always takes a first *context*
parameter (named ``ctx`` in the example above) which is an instance of
:class:`pyarrow.compute.ScalarUdfContext`.
This context exposes several useful attributes, particularly a
:attr:`~pyarrow.compute.ScalarUdfContext.memory_pool` to be used for
allocations in the context of the user-defined function.

You can call a user-defined function directly using :func:`pyarrow.compute.call_function`:

.. code-block:: python

   >>> pc.call_function("affine", [pa.scalar(2.5), pa.scalar(10.5), pa.scalar(5.5)])
   <pyarrow.DoubleScalar: 31.75>

.. note::
   Note that when the passed values to a function are all scalars, internally each scalar 
   is passed as an array of size 1.

More generally, user-defined functions are usable everywhere a compute function
can be referred to by its name. For example, they can be called on a dataset's
column using :meth:`Expression._call`:
Considering a series of scalar inputs,

.. code-block:: python

>    >>> import pyarrow as pa

Reviewer: Why is this repeating the entire example? I'm afraid I don't understand the point of this.

Author: I was merely thinking about a user who is going to just copy the codeblock and run. They are separate code blocks and if a user picks one in the middle it won’t run. Of course we can remove it. But this what I thought.

Author: I might want to add it to the one below too. But for non repetition, we can just import it once.

   >>> import pyarrow.compute as pc
   >>> function_name = "affine_with_python"
   >>> function_docs = {
   ...        "summary": "Calculate y = mx + c with Python",
   ...        "description":
   ...            "Compute the affine function y = mx + c.\n"
   ...            "This function takes three inputs, m, x and c, in order."
   ... }
   >>> input_types = {
   ...    "m" : pa.float64(),
   ...    "x" : pa.float64(),
   ...    "c" : pa.float64(),
   ... }
   >>> output_type = pa.float64()
   >>> 
   >>> def affine_with_python(ctx, m, x, c):
   ...     m = m[0].as_py()
   ...     x = x[0].as_py()
   ...     c = c[0].as_py()

>    ...     return pa.array([m * x + c])

Reviewer: Why is this taking `m[0]`? I don't understand what this example is supposed to show...

Author: I assumed you asked me to show the all scalar scenario in an example rather than wording. That’s why I added one to showcase that.

Author: I referred to this: https://github.com/apache/arrow/pull/13687#discussion_r927582170

Author: DId I misinterpret your statement?

Reviewer: Well, the all-scalar scenario works with the original UDF? no?

Author: Yes, no issue there. But the user need to know that part and how to handle it if they are using other libraries within the UDFs.

Author: @pitrou should we keep this example or just keep a note/warning? I am inclined towards the example, though.

Author: I re-organized and included content. Please check it.

Reviewer: I'm sorry, but the example still makes no sense to me. Why not use the regular affine function here, instead of this scalar-specific definition?

Reviewer 2: It is interesting, I think, that a person can define a UDF without using the Arrow compute functions at all, that is the most compelling point of the UDF feature in my mind since compositions of Arrow compute functions could already be done using expressions.

However, it is not clear from the description that this is the purpose of this example (is it?).  It's also perhaps not the most motivating example since it can be expressed as an Arrow expression.

   ... 
   >>> pc.register_scalar_function(affine_with_python,
   ...                             function_name,
   ...                         function_docs,
   ...                             input_types,
   ...                             output_type)
   >>> 
   >>> pc.call_function(function_name, [pa.scalar(10.1), pa.scalar(10.2), pa.scalar(20.2)])
   <pyarrow.DoubleScalar: 123.22>

When all the inputs are scalar, the input is a size=1 array and the values has to be properly

> treated within the UDF. And also make sure to include the final output as a size=1 array.

Reviewer: Isn't this repeating what you already said in the note above?

Author: Mmm… yes I will trim this.

UDFs can be used with tabular data by using `dataset` API and apply a UDF function on the

> dataset.

Reviewer: Here the "More generally ..." sentence should come.

Author: 👍

:class:`.Dataset` currently can be filtered using :meth:`.Dataset.to_table` method
passing a ``filter`` argument. See :ref:`py-filter-dataset` in Dataset documentation.

User-Defined Functions
======================

.. warning::
   This API is **experimental**.

PyArrow allows defining and registering custom compute functions.
These functions can then be called from Python as well as C++ (and potentially
any other implementation wrapping Arrow C++, such as the R ``arrow`` package)
using their registered function name.

To register a UDF, a function name, function docs, input types and
output type need to be defined. Using :func:`pyarrow.compute.register_scalar_function`,

.. code-block:: python

   import numpy as np

   import pyarrow as pa
   import pyarrow.compute as pc

   function_name = "numpy_gcd"
   function_docs = {
         "summary": "Calculates the greatest common divisor",
         "description":
            "Given 'x' and 'y' find the greatest number that divides\n"
            "evenly into both x and y."
   }

   input_types = {
      "x" : pa.int64(),
      "y" : pa.int64()
   }

   output_type = pa.int64()

   def to_np(val):
      if isinstance(val, pa.Scalar):

>          return val.as_py()

Reviewer 3: Is this conversion to Scalar still needed? (with the latest refactor of Wes, scalars might now be handled by len-1 arrays?) 
In any case, the `gcd_numpy` function itself won't work with scalars because of the `pa.array(..)` call in it:

```
In [32]: gcd_numpy(None, pa.scalar(27), pa.scalar(63))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-32-5dc8dd5d05b1> in <module>
----> 1 gcd_numpy(None, pa.scalar(27), pa.scalar(63))

<ipython-input-26-1579a8ef575a> in gcd_numpy(ctx, x, y)
     22    np_x = to_np(x)
     23    np_y = to_np(y)
---> 24    return pa.array(np.gcd(np_x, np_y))
     25 pc.register_scalar_function(gcd_numpy,
     26                            function_name,

~/scipy/repos/arrow/python/pyarrow/array.pxi in pyarrow.lib.array()

~/scipy/repos/arrow/python/pyarrow/array.pxi in pyarrow.lib._sequence_to_array()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.pyarrow_internal_check_status()

TypeError: 'numpy.int64' object is not iterable
```

Author: @jorisvandenbossche if we do that, we get the following error

```bash
>>> pc.call_function("numpy_gcd", [pa.scalar(27), pa.array([81, 12, 5])])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/_compute.pyx", line 560, in pyarrow._compute.call_function
  File "pyarrow/_compute.pyx", line 355, in pyarrow._compute.Function.call
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/_compute.pyx", line 2506, in pyarrow._compute._scalar_udf_callback
  File "<stdin>", line 4, in gcd_numpy
  File "/Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/numpy/core/_internal.py", line 790, in _gcd
    a, b = b, a % b
TypeError: unsupported operand type(s) for %: 'pyarrow.lib.Int64Scalar' and 'int'
```

I think the reason is, Numpy cannot identify the passed in Arrow scalar value. We need to take the python value of it or convert it to numpy. 

The following is what would happen 

```bash
>>> np.gcd(np.array(pa.scalar(27)), np.array([81, 12, 5]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/numpy/core/_internal.py", line 790, in _gcd
    a, b = b, a % b
TypeError: unsupported operand type(s) for %: 'pyarrow.lib.Int64Scalar' and 'int'
```

But it would work for 

```bash
np.gcd(np.array(pa.array([27])), np.array([81, 12, 5]))
array([27,  3,  1])
```

but not for 

```bash
>>> np.gcd(np.array(pa.scalar(27)), np.array([81, 12, 5]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/vibhatha/venv/pyarrow_dev/lib/python3.10/site-packages/numpy/core/_internal.py", line 790, in _gcd
    a, b = b, a % b
TypeError: unsupported operand type(s) for %: 'pyarrow.lib.Int64Scalar' and 'int'
```

And again works for

```bash
>>> np.gcd(np.array(27), np.array([81, 12, 5]))
array([27,  3,  1])
```

Am I wrong here or missing something?

Reviewer 3: OK, so it seems that we are _still_ passing scalars to the UDF in case of mixed scalar / array arguments, like you do in `pc.call_function("numpy_gcd", [pa.scalar(27), pa.array([81, 12, 5])])`. 

So my understanding was wrong based on testing the case of two scalars: `pc.call_function("numpy_gcd", [pa.scalar(27), pa.scalar(81)])`. That actually works by passing two length-1 arrays to the UDF. 
(and for that reason I assumed we always pass scalars as length-1 arrays to the UDF implementation)

Personally, I find this a bit confusing (but I am not familiar enough with the internals of the kernels to know if it is easy to always pass arrays to the UDF). And at least I would document that the arguments can be scalars, but never all of them will be scalar at the same time. 
(since as I mentioned above, the current example UDF implementation doesn't work if it would be passed only scalars)

Author: I understand your point. That's why the description included that idea. We didn't change the UDF kernels since it was merged. So for the moment I will document what's missing from the current docs. cc @westonpace @pitrou 
What's the best?

      else:
         return np.array(val)

   def gcd_numpy(ctx, x, y):

>       np_x = to_np(x)

Reviewer 3: Small nitpick: can you use 4-space indentation in the python snippets?

      np_y = to_np(y)
      return pa.array(np.gcd(np_x, np_y))

   pc.register_scalar_function(gcd_numpy,
                              function_name,
                              function_docs,
                              input_types,
                              output_type)
   

The implementation of a user-defined function always takes first *context*
parameter (named ``ctx`` in the example above) which is an instance of
:class:`pyarrow.compute.ScalarUdfContext`.
This context exposes several useful attributes, particularly a
:attr:`~pyarrow.compute.ScalarUdfContext.memory_pool` to be used for
allocations in the context of the user-defined function.

PyArrow UDFs accept input types of both :class:`~pyarrow.Scalar` and :class:`~pyarrow.Array`,
and there will always be at least one input of type :class:`~pyarrow.Array`.

> The output should always be a :class:`~pyarrow.Array`.

Reviewer 3: This is probably related to my comment above, and could indeed explain why the function would work with scalars (if one of both is not an array, numpy will also return an array, and the `pa.array(..)` call will not fail)

However, adding a print statement for the type of argument in `to_np` and running this example again, I see:

```
In [3]: pc.call_function("numpy_gcd", [pa.scalar(27), pa.scalar(63)])
<class 'pyarrow.lib.Int64Array'>
<class 'pyarrow.lib.Int64Array'>
Out[3]: <pyarrow.Int64Scalar: 9>
```

So it seems that both arguments are converted to an array. If that is guaranteed to always be the case now, the above paragraph is outdated.

Author: Ah nice catch on the description. Removing it.

You can call a user-defined function directly using :func:`pyarrow.compute.call_function`:

.. code-block:: python

   >>> pc.call_function("numpy_gcd", [pa.scalar(27), pa.scalar(63)])
   <pyarrow.Int64Scalar: 9>
   >>> pc.call_function("numpy_gcd", [pa.scalar(27), pa.array([81, 12, 5])])
   <pyarrow.lib.Int64Array object at 0x7fcfa0e7b100>
   [
     27,
     3,
     1
   ]

Working with Datasets
---------------------

More generally, user-defined functions are usable everywhere a compute function
can be referred to by its name. For example, they can be called on a dataset's
column using :meth:`Expression._call`.

Consider an instance where the data is in a table and we want to compute
the GCD of one column with the scalar value 30.  We will be re-using the
"numpy_gcd" user-defined function that was created above:

.. code-block:: python

   >>> import pyarrow.dataset as ds

>    >>> sample_data = {'category': ['A', 'B', 'C', 'D'], 'value': [90, 630, 1827, 2709]}
>    >>> data_table = pa.Table.from_pydict(sample_data)

Reviewer 3: ```suggestion
   >>> data_table = pa.table({'category': ['A', 'B', 'C', 'D'], 'value': [90, 630, 1827, 2709]})
```

(a bit simpler to create the same table)

:class:`.Dataset` currently can be filtered using :meth:`.Dataset.to_table` method
passing a ``filter`` argument. See :ref:`py-filter-dataset` in Dataset documentation.

User-Defined Functions
======================

.. warning::
   This API is **experimental**.

PyArrow allows defining and registering custom compute functions.
These functions can then be called from Python as well as C++ (and potentially
any other implementation wrapping Arrow C++, such as the R ``arrow`` package)
using their registered function name. 

> UDF support is limited to scalar functions. A scalar function is a function which
> executes elementwise operations on arrays or scalars. In general, the output of a
> scalar function do not depend on the order of values in the arguments. Note that 
> such functions have a rough correspondence to the functions used in SQL expressions.

Reviewer: ```suggestion
UDF support is limited to scalar functions. A scalar function is a function which
executes elementwise operations on arrays or scalars. In general, the output of a
scalar function does not depend on the order of values in the arguments. Note that 
such functions have a rough correspondence to the functions used in SQL expressions,
or to NumPy `universal functions <https://numpy.org/doc/stable/reference/ufuncs.html>`_.
```

To register a UDF, a function name, function docs, input types and
output type need to be defined. Using :func:`pyarrow.compute.register_scalar_function`,

.. code-block:: python

   import numpy as np

   import pyarrow as pa
   import pyarrow.compute as pc

   function_name = "numpy_gcd"
   function_docs = {
         "summary": "Calculates the greatest common divisor",
         "description":
            "Given 'x' and 'y' find the greatest number that divides\n"
            "evenly into both x and y."
   }

   input_types = {
      "x" : pa.int64(),
      "y" : pa.int64()
   }

   output_type = pa.int64()

   def to_np(val):
       if isinstance(val, pa.Scalar):
          return val.as_py()
       else:
          return np.array(val)

   def gcd_numpy(ctx, x, y):
       np_x = to_np(x)
       np_y = to_np(y)
       return pa.array(np.gcd(np_x, np_y))

   pc.register_scalar_function(gcd_numpy,
                              function_name,
                              function_docs,
                              input_types,
                              output_type)
   

> The implementation of a user-defined function always takes first *context*
> parameter (named ``ctx`` in the example above) which is an instance of
> :class:`pyarrow.compute.ScalarUdfContext`.

Reviewer: ```suggestion
The implementation of a user-defined function always takes a first *context*
parameter (named ``ctx`` in the example above) which is an instance of
:class:`pyarrow.compute.ScalarUdfContext`.
```

and there will always be at least one input of type :class:`~pyarrow.Array`.
The output should always be a :class:`~pyarrow.Array`.

You can call a user-defined function directly using :func:`pyarrow.compute.call_function`:

.. code-block:: python

   >>> pc.call_function("numpy_gcd", [pa.scalar(27), pa.scalar(63)])
   <pyarrow.Int64Scalar: 9>
   >>> pc.call_function("numpy_gcd", [pa.scalar(27), pa.array([81, 12, 5])])
   <pyarrow.lib.Int64Array object at 0x7fcfa0e7b100>
   [
     27,
     3,
     1
   ]

Working with Datasets
---------------------

More generally, user-defined functions are usable everywhere a compute function

> can be referred by its name. For example, they can be called on a dataset's

Reviewer 4: ```suggestion
can be referred to by its name. For example, they can be called on a dataset's
```

"numpy_gcd" user-defined function that was created above:

.. code-block:: python

   >>> import pyarrow.dataset as ds
   >>> sample_data = {'category': ['A', 'B', 'C', 'D'], 'value': [90, 630, 1827, 2709]}
   >>> data_table = pa.Table.from_pydict(sample_data)
   >>> dataset = ds.dataset(data_table)
   >>> func_args = [pc.scalar(30), ds.field("value")]
   >>> dataset.to_table(
   ...             columns={
   ...                 'gcd_value': ds.field('')._call("numpy_gcd", func_args),
   ...                 'value': ds.field('value'),
   ...                 'category': ds.field('category')
   ...             })
   pyarrow.Table
   gcd_value: int64
   value: int64
   category: string
   ----
   gcd_value: [[30,30,3,3]]
   value: [[90,630,1827,2709]]
   category: [["A","B","C","D"]]

Note that ``ds.field('')_call(...)`` returns a :func:`pyarrow.compute.Expression`.
The arguments passed to this function call are expressions, not scalar values 
(notice the difference between :func:`pyarrow.scalar` and :func:`pyarrow.compute.scalar`,
the latter produces an expression). 
This expression is evaluated when the projection operator executes it.

Projection Expressions
^^^^^^^^^^^^^^^^^^^^^^

> In the above example we used an expression to add a new column (``gcd_value``)

Author: Ah not exactly. But it was added as a sub section to give clarity on what projection expression does and how it can be used with UDFs.

> In the above example we used an expression to add a new column (``gcd_value``)

Reviewer 3: I think we should also mention somewhere (more to the beginning of the new section, I think), that currently the UDFs are limited to scalar functions (and then also explain what a scalar function is). 
That will also make it easier to refer to that concept here to say that projections currently only support scalar functions.

> In the above example we used an expression to add a new column (``gcd_value``)

Author: Yeah that make sense. Will update.

> In the above example we used an expression to add a new column (``gcd_value``)

Author: I added a section the limitations and scalar function definition. I merely rephrased what is the docs to be precise. Didn't want introduce anything extra.

   function_docs = {
         "summary": "Calculates the greatest common divisor",
         "description":
            "Given 'x' and 'y' find the greatest number that divides\n"
            "evenly into both x and y."
   }

   input_types = {
      "x" : pa.int64(),
      "y" : pa.int64()
   }

   output_type = pa.int64()

   def to_np(val):
      if isinstance(val, pa.Scalar):
         return val.as_py()
      else:
         return np.array(val)

   def gcd_numpy(ctx, x, y):
      np_x = to_np(x)
      np_y = to_np(y)
      return pa.array(np.gcd(np_x, np_y))

   pc.register_scalar_function(gcd_numpy,
                              function_name,
                              function_docs,
                              input_types,
                              output_type)
   

The implementation of a user-defined function always takes first *context*
parameter (named ``ctx`` in the example above) which is an instance of
:class:`pyarrow.compute.ScalarUdfContext`.
This context exposes several useful attributes, particularly a
:attr:`~pyarrow.compute.ScalarUdfContext.memory_pool` to be used for
allocations in the context of the user-defined function.

> PyArrow UDFs accept input types of both scalar and array. Also it can have
> any combination of these types. It is important that the UDF author ensures
> the UDF can handle such combinations correctly. Also the ability to use UDFs
> with existing data processing libraries is very useful.

Reviewer: ```suggestion
PyArrow UDFs accept input types of both :class:`~pyarrow.Scalar` and :class:`~pyarrow.Array`,
and there will always be at least one input of type :class:`~pyarrow.Array`.
The output should always be a :class:`~pyarrow.Array`.
```

> Note that there is a helper function `to_np` to handle the conversion 

Reviewer: This paragraph doesn't look useful to me, just remove it?

Author: I am removing the section after `Note that there is ...`

of scalar and array inputs to the UDF. Also, the final output is returned
as a scalr or an array depending on the inputs. Based on the usage of any
libraries inside the UDF, make sure it is generalized to support the passed
input values and return suitable values.

You can call a user-defined function directly using :func:`pyarrow.compute.call_function`:

.. code-block:: python

   >>> pc.call_function(""numpy_gcd", [pa.scalar(27), pa.scalar(63)])
   9

Working with Datasets
---------------------

More generally, user-defined functions are usable everywhere a compute function
can be referred by its name. For example, they can be called on a dataset's

> column using :meth:`Expression._call`:

Reviewer: ```suggestion
column using :meth:`Expression._call`.
```

> Consider an instance where the data is in a table and you need to create a new 
> column using existing values in another column by using a mathematical formula.
> For instance, let's consider applying `gcd` math operation.
> Here, we will be re-using the registered `numpy_gcd` function.

Reviewer: ```suggestion
Consider an instance where the data is in a table and we want to compute
the GCD of one column with the scalar value 30.  We will be re-using the
"numpy_gcd" user-defined function that was created above:
```

.. code-block:: python

   >>> import pyarrow.dataset as ds
   >>> sample_data = {'category': ['A', 'B', 'C', 'D'], 'value': [90, 630, 1827, 2709]}
   >>> data_table = pa.Table.from_pydict(sample_data)
   >>> dataset = ds.dataset(data_table)
   >>> func_args = [pc.scalar(30), ds.field("value")]
   >>> dataset.to_table(
   ...             columns={
   ...                 'gcd_value': ds.field('')._call("numpy_gcd", func_args),
   ...                 'value': ds.field('value'),
   ...                 'category': ds.field('category')
   ...             })
   pyarrow.Table
   gcd_value: int64
   value: int64
   category: string
   ----
   gcd_value: [[30,30,3,3]]
   value: [[90,630,1827,2709]]
   category: [["A","B","C","D"]]

Note that the `ds.field('')_call()` returns an expression. The passed arguments
to this function call are expressions not scalar values 
(i.e `pc.scalar(30), ds.field("value")`, notice the difference 
of `pa.scalar` vs `pc.scalar`, the latter produces an expression). 
This expression is evaluated when the project operator executes it.

Projection Expressions
^^^^^^^^^^^^^^^^^^^^^^

> In the above example we used an expression to add a new column (`gcd_value`)

Reviewer: ```suggestion
In the above example we used an expression to add a new column (``gcd_value``)
```

to our table.  Adding new, dynamically computed, columns to a table is known as "projection"
and there are limitations on what kinds of functions can be used in projection expressions.
A projection function must emit a single output value for each input row.  That output value
should be calculated entirely from the input row and should not depend on any other row.
For example, the "numpy_gcd" function that we've been using as an example above is a valid
function to use in a projection.  A "cumulative sum" function would not be a valid function

> since the result of each input rows depends on the rows that came before.  A "drop nulls"

Reviewer: ```suggestion
since the result of each input row depends on the rows that came before.  A "drop nulls"
```

      "x" : pa.float64(),
      "c" : pa.float64(),
   }
   output_type = pa.float64()

   def affine(ctx, m, x, c):
       temp = pc.multiply(m, x, memory_pool=ctx.memory_pool)
       return pc.add(temp, c, memory_pool=ctx.memory_pool)

   pc.register_scalar_function(affine, 
                               function_name,
                               function_docs,
                               input_types,
                               output_type)

The implementation of a user-defined function always takes a first *context*
parameter (named ``ctx`` in the example above) which is an instance of
:class:`pyarrow.compute.ScalarUdfContext`.
This context exposes several useful attributes, particularly a
:attr:`~pyarrow.compute.ScalarUdfContext.memory_pool` to be used for
allocations in the context of the user-defined function.

You can call a user-defined function directly using :func:`pyarrow.compute.call_function`:

.. code-block:: python

   >>> pc.call_function("affine", [pa.scalar(2.5), pa.scalar(10.5), pa.scalar(5.5)])
   <pyarrow.DoubleScalar: 31.75>

Generalizing Usage
------------------

PyArrow UDFs accept input types of both scalar and array. Also it can have

> vivid combinations of these types. It is important that the UDF author must make sure,

Reviewer: Is "vivid" the right word here? @westonpace What do you think?

Reviewer 2: I think `any` would be better.

the UDF is defined such that it can handle such combinations well. 

For instance, when the passed values to a function are all scalars, internally
each scalar is passed as an array of size 1.

To elaborate on this, let's consider a scenario where we have a function
which computes a scalar `y` value based on scalar inputs 
`m`, `x` and `c` using python arithmetic operations.

.. code-block:: python

   >>> import pyarrow.compute as pc
   >>> function_name = "affine_with_python"
   >>> function_docs = {
   ...        "summary": "Calculate y = mx + c with Python",
   ...        "description":
   ...            "Compute the affine function y = mx + c.\n"
   ...            "This function takes three inputs, m, x and c, in order."
   ... }
   >>> input_types = {
   ...    "m" : pa.float64(),
   ...    "x" : pa.float64(),
   ...    "c" : pa.float64(),
   ... }
   >>> output_type = pa.float64()
   >>> 
   >>> def affine_with_python(ctx, m, x, c):
   ...     m = m[0].as_py()
   ...     x = x[0].as_py()
   ...     c = c[0].as_py()
   ...     return pa.array([m * x + c])
   ... 
   >>> pc.register_scalar_function(affine_with_python,
   ...                             function_name,

>    ...                         function_docs,

Reviewer 2: ```suggestion
   ...                             function_docs,
```

   ...                             input_types,
   ...                             output_type)
   >>> 
   >>> pc.call_function(function_name, [pa.scalar(10.1), pa.scalar(10.2), pa.scalar(20.2)])
   <pyarrow.DoubleScalar: 123.22>

> Note that here the the final output is returned as an array. Depending the usage of vivid libraries

Reviewer 2: I'm not sure what you are trying to say with the sentence that starts "Depending the..."

mathematical expression, `y = mx + c`. We will be re-using the registered `affine`
function.

.. code-block:: python

   >>> import pyarrow.dataset as ds
   >>> sample_data = {'category': ['A', 'B', 'C', 'D'], 'value': [10.21, 20.12, 45.32, 15.12]}
   >>> data_table = pa.Table.from_pydict(sample_data)
   >>> dataset = ds.dataset(data_table)
   >>> func_args = [pc.scalar(5.2), ds.field("value"), pc.scalar(2.1)]
   >>> dataset.to_table(
   ...             columns={
   ...                 'projected_value': ds.field('')._call("affine", func_args),
   ...                 'value': ds.field('value'),
   ...                 'category': ds.field('category')
   ...             })
   pyarrow.Table
   total_amount_projected($): int64
   total_amount($): int64
   trip_name: string
   ----
   total_amount_projected($): [[52,102,227,77]]
   total_amount($): [[10,20,45,15]]
   trip_name: [["A","B","C","D"]]

Here note that the `ds.field('')_call()` returns an expression. The passed arguments
to this function call are expressions not scalar values 
(i.e `pc.scalar(5.2), ds.field("value"), pc.scalar(2.1)`). This expression is evaluated

> when the project operator uses this expression.

Reviewer 2: You say "The passed arguments to this function call are expressions not scalar values".

However, `pc.scalar(5.2)` and `pc.scalar(2.1)` look like scalar values.  I'm not sure a user will recognize the subtle difference between `pc.scalar(5.2)` and `pa.scalar(5.2)` without further explanation.

> Support
> -------
> 
> It is defined that the current support is only for scalar functions. 
> A scalar function (:class:`arrow::compute::ScalarFunction`) executes elementwise operations
> on arrays or scalars. Generally, the result of such an execution doesn't
> depend on the order of values.
> 
> There is a limitation in the support to UDFs in the current API.
> For instance, with project node, if a UDF is used as the compute function,
> it expects the function to be a scalar function. Although, this doesn't stop the user
> registering a non-scalar function and using it in a programme. 
> But it could lead to unexpected behaviors or errors when it is applied in such occasions. 
> The current UDF support could enhance with the addition of more settings to the API (i.e aggregate UDFs).

Reviewer 2: ```suggestion
Projection Expressions
^^^^^^^^^^^^^^^^^^^^^^

In the above example we used an expression to add a new column (`total_amount_projected`)
to our table.  Adding new, dynamically computed, columns to a table is known as "projection"
and there are limitations on what kinds of functions can be used in projection expressions.

A projection function must emit a single output value for each input row.  That output value
should be calculated entirely from the input row and should not depend on any other row.
For example, the "affine" function that we've been using as an example above is a valid
function to use in a projection.  A "cumulative sum" function would not be a valid function
since the result of each input rows depends on the rows that came before.  A "drop nulls"
function would also be invalid because it doesn't emit a value for some rows.
```

   ... }
   >>> input_types = {
   ...    "m" : pa.float64(),
   ...    "x" : pa.float64(),
   ...    "c" : pa.float64(),
   ... }
   >>> output_type = pa.float64()
   >>> 
   >>> def to_numpy(val):
   ...     if isinstance(val, pa.Scalar):
   ...         return val.as_py()
   ...     else:
   ...         return np.array(val)
   ... 
   >>> def affine_with_numpy(ctx, m, x, c):
   ...     m = to_numpy(m)
   ...     x = to_numpy(x)
   ...     c = to_numpy(c)
   ...     return pa.array(m * x + c)
   ... 
   >>> pc.register_scalar_function(affine_with_numpy,
   ...                             function_name,
   ...                             function_docs,
   ...                             input_types,
   ...                             output_type)
   >>> pc.call_function(function_name, [pa.scalar(10.1), pa.scalar(10.2), pa.scalar(20.2)])
   <pyarrow.DoubleScalar: 123.22>
   >>> pc.call_function(function_name, [pa.scalar(10.1), pa.array([10.2, 20.2]), pa.scalar(20.2)])
   <pyarrow.lib.DoubleArray object at 0x10e38eb20>
   [
      123.22,
      224.21999999999997
   ]

Note that there is a helper function `to_numpy` to handle the conversion of scalar an array inputs

> to the UDf. Also, the final output is returned as a scalr or an array depending on the inputs.

Reviewer 3: ```suggestion
to the UDF. Also, the final output is returned as a scalar or an array depending on the inputs.
```

:class:`.Dataset` currently can be filtered using :meth:`.Dataset.to_table` method
passing a ``filter`` argument. See :ref:`py-filter-dataset` in Dataset documentation.

User-Defined Functions
======================

.. warning::
   User-defined functions only supports scalar functions and the current version is experimental.

To use a user-defined-function (UDF), either the experimental `dataset` API options can be used or the
function can be directly called using :func:`pyarrow.compute.call_function`. 

To register a UDF, a function name, function docs and input types and output type needs to be defined.

.. code-block:: python

   import pyarrow.compute as pc

>    function_name = "regression"

Author: Ah I think I started thinking about writing a regression and ended up with a simple equation. Let's rename it to something like `cost_update`?

>    function_name = "regression"

Reviewer: Why cost_update? What does it have to do with costs?

>    function_name = "regression"

Author: may be `project_price`

>    function_name = "regression"

Author: Or `y=mx+b` simply.

>    function_name = "regression"

Reviewer: "affine" perhaps?

>    function_name = "regression"

Author: Sure, that's much better.

   function_docs = {
      "summary": "Calculate y based on m, x and c values",
      "description": "Obtaining output of a linear scalar function"
   }
   input_types = {
      "m" : pa.int64(),
      "x" : pa.int64(),
      "c" : pa.int64(),
   }

>    output_type = pa.int64()

Author: Yeah we should probably get floats to this.

>    def linear_calculation(ctx, m, x, c):
>       return pc.add(pc.multiply(m, x), c)

Reviewer: Can we perhaps use the adequate memory pool here? For example:
```suggestion
   def affine_calculation(ctx, m, x, c):
       temp = pc.multiply(m, x, memory_pool=ctx.memory_pool)
       return pc.add(temp, c, memory_pool=ctx.memory_pool)
```

>    pc.register_scalar_function(linear_calculation, 

Reviewer: It's not linear but affine, so should fix the name.

                               function_name,
                               function_docs,
                               input_types,
                               output_type)

> .. note::
>    There is a default parameter, `ctx` which is refers to a context object and it should be the
>    first parameter of any user-defined-function. The idea is to make available passing required
>    meta-data across an application which would be important for UDFs.

Reviewer: ```suggestion
The implementation of a user-defined function always takes a first *context*
parameter (named ``ctx`` in the example above) which is an instance of
:class:`pyarrow.compute.ScalarUdfContext`.
This context exposes several useful attributes, particularly a
:attr:`~pyarrow.compute.ScalarUdfContext.memory_pool` to be used for
allocations in the context of the user-defined function.
```

Calling a UDF directly using :func:`pyarrow.compute.call_function`,

.. code-block:: python

   >>> res = pc.call_function("regression", [pa.scalar(2), pa.scalar(10), pa.scalar(5)])
   25

.. warning::
   Note that when the passed values to a function are all scalars, internally each scalar 
   is passed as an array of size 1.

> UDFs can be used with tabular data by using `dataset` API and apply a UDF function on the
> dataset.

Reviewer: Perhaps rephrase it to make it less exceptional:
```suggestion
More generally, user-defined functions are usable everywhere a compute function
can be referred to by its name. For example, they can be called on a dataset's
column using :meth:`Expression._call`:
```

.. code-block:: python

   >>> sample_data = {'trip_name': ['A', 'B', 'C', 'D'], 'total_amount($)': [10, 20, 45, 15]}
   >>> data_table = pa.Table.from_pydict(sample_data)

>    >>> import pyarrow.dataset as ds

Reviewer: Please move the import at the beginning of this snippet.

   >>> dataset = ds.dataset(data_table)
   >>> func_args = [pc.scalar(5), ds.field("total_amount($)"), pc.scalar(2)]
   >>> result_table = dataset.to_table(
   ...             columns={

>    ...                 'total_amount_projected($)': ds.field('')._call(function_name, func_args),

Reviewer: What is `function_name` here? Perhaps it would be better to spell it explicitly.

Reviewer: Also @jorisvandenbossche do you remember why `Expression._call` has a leading underscore?

Author: `function_name` replaced with "affine"
